core/txpool/legacypool/legacypool_test.go

```diff
@@ -1167,28 +1167,62 @@ func TestAllowedTxSize(t *testing.T) {
 	account := crypto.PubkeyToAddress(key.PublicKey)
 	testAddBalance(pool, account, big.NewInt(1000000000))
 
-	// Find the maximum data length for the kind of transaction which will
-	// be generated in the pool.addRemoteSync calls below.
-	const largeDataLength = txMaxSize - 200 // enough to have a 5 bytes RLP encoding of the data length number
-	txWithLargeData := pricedDataTransaction(0, pool.currentHead.Load().GasLimit, big.NewInt(1), key, largeDataLength)
-	maxTxLengthWithoutData := txWithLargeData.Size() - largeDataLength // 103 bytes
-	maxTxDataLength := txMaxSize - maxTxLengthWithoutData              // 131072 - 103 = 130953 bytes
+	// Find the "usual" transaction length without data.
+	// This varies depending on the data length and data content in the transaction:
+	// - The data length encoding varies from 0 to 5 bytes
+	// - The signature encoding varies from 3 bytes to 68 bytes depending on the data content
+	// However, in most cases, the maximum transaction length without data is 103 bytes,
+	// which is used as a starting point to create transactions below.
+	txWithLargeData := pricedDataTransaction(0, pool.currentHead.Load().GasLimit, big.NewInt(1), key, txMaxSize)
+	usualTxLengthWithoutData := txWithLargeData.Size() - txMaxSize
+
+	makeTx := func(t *testing.T, nonce, targetSize uint64) *types.Transaction {
+		t.Helper()
+		gasLimit := pool.currentHead.Load().GasLimit
+		gasPrice := big.NewInt(1)
+
+		dataLength := targetSize - usualTxLengthWithoutData
+
+		const tries = 68 - // 68 is the maximum signature encoding size
+			3 + // 3 is the minimum signature encoding size
+			5 - // 5 is the maximum data length encoding size
+			0 // 0 is the minimum data length encoding size
+		for range tries {
```

Comment:
I think this could be simplified by assuming that we want to find a target signature size of 65 bytes, and then in an infinite loop generate signed data transactions of `targetSize -65` and  return if the signed tx size is the target size.

Because in practice, it's guaranteed to terminate quickly.

Reply:
I'm still working on this, it's taking a bit longer than expected, please keep this opened ~24 hours and I'll get it done. Thanks!

Reply:
Thanks for taking the time to make these changes.  I think I'm only now realizing that creating a generic `sizedDataTransaction` function is pretty non-trivial.

Given that the test-case is really concerned about the boundary conditions (`maxSize - 1`, `maxSize`, `maxSize + 1`), I'm wondering if we could pair down the complexity of this test by only generating txs of these sizes.

It's not quite as nice as generating a wider variety of inputs, but it could probably suffice here.

Reply:
If we have a fixed key and deterministic data generation (no crand.Read), there are two ways to do this:

1. have fully pre-determined transactions, for example as hex encoding of RLP encoding. This removes a lot of logic, but makes the source code rather big given transactions are rather big in sizes. We could move these to a testdata directory. Thoughts?
2. Keep generating with a bit less logic, given the signature length can vary with the data length, even if the key is fixed and the data is all zeroes. This is what I've done in the last commit 399949cfa2d5307812a574b715936b9813266bdf